File: libmonet/lib/palettes/tonal_palette.js
'use strict';
import { HCT } from '../hct.js';

export class TonalPalette {
    constructor(hue, chroma) {
        this.hue = hue;
        this.chroma = chroma;
        this.cache = new Map();
    }
    tone(tone) {
        let argb = putIfAbsent(this.cache, tone, _ => new HCT(this.hue, this.chroma, tone).toInt());
        return argb;
    }
};

function putIfAbsent(map, key, create) {
    if (map.has(key)) {
        return map.get(key)
    }
    const val = create();
    map.set(key, val);
    return val;
}

File: dart/libmonet/lib/palettes/tonal_palette.dart
import 'package:libmonet/hct/hct.dart';

class TonalPalette {
  final num hue;
  final num chroma;
  final Map<int, int> cache = {};

  TonalPalette(this.hue, this.chroma);
  int tone(int tone) {
    final argb = cache.putIfAbsent(
      tone,
      () => HCT(hue, chroma, tone).toInt(),
    );
    return argb;
  }
}
